PR pour correction du code

sukivana opened the pull request:
Bonjour, je pense que 2 points par semaine me conviendrait

Commit:
Ajout de la gestion de "Personne". 

Param
Vue avec recueil des données
2 "models" affichage et ajout

Commit:
Première sauvegarde

Commit:
Version du 6/12/16

Commit:
suite 2 du 6/12/16

Commit:
Merge branch 'MVC_Person'

Conflicts:
	etape1/src/main/java/org/iae/annecy/st1/etape1/Main.java
	etape1/src/main/java/org/iae/annecy/st1/tools/ConsoleHelper.java

Commit:
Commit du 14/12

Commit:
14/12

Commit:
Presque derniere version

Commit:
Code version finale

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/controller/BaseDeDonneesController.java:
```diff
@@ -0,0 +1,55 @@
+public class BaseDeDonneesController implements Serializable{
+
+    BaseDeDonnees bdd = new BaseDeDonnees();
+    BaseDeDonneesView bddView = new BaseDeDonneesView(this);
```
(majeur) Semble inutilisé, a supprimer.

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/controller/BaseDeDonneesController.java:
```diff
@@ -0,0 +1,55 @@
+@SuppressWarnings("serial")
+public class BaseDeDonneesController implements Serializable{
+
+    BaseDeDonnees bdd = new BaseDeDonnees();
```
(majeur) Ajouter "private" devant, par defaut les attributs ont la même visibilité que la classe (donc public dans ce cas). Pour suivre le principe d'encapsulation (POO) il FAUT passer par les getter/setter pour modifier cette variable.

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/controller/BaseDeDonneesController.java:
```diff
@@ -0,0 +1,55 @@
+import org.iae.annecy.st1.etape1.view.BaseDeDonneesView;
+
+@SuppressWarnings("serial")
+public class BaseDeDonneesController implements Serializable{
```
(majeur) Cette classe ne semble pas se comporter comme un controller (on aborde MVC un peu plus tard). Cette classe semble être un "doublons" de ta classe org.iae.annecy.st1.etape1.model.person.BaseDeDonnees. Je te propose de supprimer cette classe.

- Met a jour ta méthode org.iae.annecy.st1.etape1.model.person.BaseDeDonnees.retrieveClient(String, String) pour prendre ne compte le renvoie systématique d'une "Person".
- Modifie les référence __org.iae.annecy.st1.etape1.controller.BaseDeDonneesController.retrouveClient(String, String)__" vers qui maintenant peuvent aller directement vers __BaseDeDonnees.retrieveClient(String, String)__ (dans Eclipse, sélectionner la methode __BaseDeDonneesController.retrouveClient(String, String)__ puis fait un clique droit -> reference -> projet. Eclipse n’indiquera partout ou cette méthode est appelées, et donc là ou tu devra faire des modifications.

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/controller/CatalogueController.java:
```diff
@@ -0,0 +1,41 @@
+import org.iae.annecy.st1.etape1.view.CatalogueView;
+
+@SuppressWarnings("serial")
+public class CatalogueController implements java.io.Serializable{
```
Idem que pour "BaseDeDonneesController". Je te propose de supprimer en modifiant les référence vers __org.iae.annecy.st1.etape1.controller.CatalogueController.sauvegarder()__ par des appels vers __org.iae.annecy.st1.etape1.model.produit.Catalogue.save()__. Pour la méthoe "get" appel directement le "Catalogue"

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/controller/PanierController.java:
```diff
@@ -0,0 +1,30 @@
+import org.iae.annecy.st1.etape1.model.panier.Panier;
+import org.iae.annecy.st1.etape1.view.PanierView;
+
+public class PanierController{
```
Idem que pour "BaseDeDonneesController". Je te propose de supprimer en modifiant les référence vers le "Get" en appelant directement le PanierView

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/model/panier/Panier.java:
```diff
@@ -0,0 +1,39 @@
+import org.iae.annecy.st1.etape1.model.person.Person;
+import org.iae.annecy.st1.etape1.model.produit.Produit;
+
+@SuppressWarnings("serial")
```
(majeur) eviter cette annotation (qui "masque" l'avertissement). Eclipse te propose aussi de "genérer un SerailVersionUID" c'est une bien meilleur option. Supprime cette ligne et dans la "correction automatique" d'Eclipse choisie "generate SerialVersionUID". Si tu veux plus d'info sur ce sujet, répond a ce commentaire en me demandant ce que tu voudrais que j’éclaircisse (quelques info synthétiques sur le sujet pour info : http://stackoverflow.com/a/285809)

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/model/person/BaseDeDonnees.java:
```diff
@@ -0,0 +1,89 @@
+    /**
+     * Sérialise la base de données clients
+     */
+    @SuppressWarnings("resource")
```
Evite cette annotation (supprime la). Pour corriger le warning "*Resource leak: 'oos' is never closed*", ajoute un block "finaly" dans ton "try/catch" ou tu ajoute un `oos.close();`. Sans, tu garde ouvert un "pointeur" vers la ressource, et à chaque __saveBdd__ tu créer un nouveau pointeur. "Heureusement" pour toi Java fait le ménage derrière toi quand tu "quite" l'application, Sinon tu devrait re-démarer TOUT ton ordinateur régulièrement pour éviter de saturer la mémoire de ton PC. Quelques infos sur le tray/catch/finally : http://tutorials.jenkov.com/java-exception-handling/basic-try-catch-finally.html#finally). N'hésite pas me demander en commentaire sur tu veux plus d'informations.

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/model/person/BaseDeDonnees.java:
```diff
@@ -0,0 +1,89 @@
+import java.util.ArrayList;
+import java.util.Iterator;
+
+@SuppressWarnings("serial")
```
Même remarque que sur org.iae.annecy.st1.etape1.model.panier.Pannier.

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/model/person/Person.java:
```diff
@@ -0,0 +1,99 @@
+
+import org.iae.annecy.st1.etape1.model.produit.Produit;
+
+@SuppressWarnings("serial")
```
Même remarque que sur org.iae.annecy.st1.etape1.model.panier.Pannier.

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/model/person/Person.java:
```diff
@@ -0,0 +1,99 @@
+     * 
+     * @return Un nombre (double)
+     */
+    public double montantPanier(){
```
Utile "__D__ouble" plutot que "__d__ouble". La version avec "D" majuscule est "Objet" donc plus orienté POO. Il n'y a que quelques très rare bonne raison d'utiliser le primitif (avec le "d" minuscule). Pour info Java à un mécanisme "d'auto-boxing" (https://docs.oracle.com/javase/tutorial/java/data/autoboxing.html) qui convertie automatiquement pour toi les primitif vers leur correspondance Objet si nécessaire (et inversement), mais autant éviter de faire-aire à Java du travail "inutile".

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/model/produit/Catalogue.java:
```diff
@@ -0,0 +1,65 @@
+import java.util.ArrayList;
+import java.util.Iterator;
+
+@SuppressWarnings("serial")
```
Même remarque que sur org.iae.annecy.st1.etape1.model.panier.Pannier.

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/model/produit/Catalogue.java:
```diff
@@ -0,0 +1,65 @@
+    /**
+     * Sérialise le catalogue
+     */
+    @SuppressWarnings("resource")
```
Même remarquey que sur org.iae.annecy.st1.etape1.model.person.saveBdd.saveBdd

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/model/produit/Produit.java:
```diff
@@ -0,0 +1,74 @@
+package org.iae.annecy.st1.etape1.model.produit;
+
+@SuppressWarnings("serial")
```
Même remarque que sur org.iae.annecy.st1.etape1.model.panier.Pannier.

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/model/produit/Produit.java:
```diff
@@ -0,0 +1,74 @@
+    public Produit(){	
+    }
+
+    public Produit(String reference,String nom, String description, double prix){
```
Préfère les "__D__ouble" au "__d__ouble"

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/model/produit/Produit.java:
```diff
@@ -0,0 +1,74 @@
+	this.prix = prix;
+    }
+
+    public Produit(String reference,String nom, String description, String descL, double prix){
```
Préfère les "__D__ouble" au "__d__ouble"

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/model/produit/Produit.java:
```diff
@@ -0,0 +1,74 @@
+
+    private String reference, descritpion, nom, descripLongue;
+    private double prix;
+    private int quantite;
```
La quantité ne peut pas être géré dans "produit", sinon, pour un produit donnée, TOUS les paniers le contenant auront la __même__ quantité. JE te propos de créer une nouvelle classe "Item" (dans xxx.xxx.xxx.model.pannier). Cette classe contiendra UN Produit et Une quantité. Ton attribut panier (org.iae.annecy.st1.etape1.model.panier.Pannier) deviendra donc une `List<Item>`. Il va falloir que tu fasse des modification un peu partout pour prendre en compte se changement. Le plus simple est de modifier ta méthode `Pannier.add(Produit p)` de la façon suivante 
```
public void add(Produit produit, Double quantite){
  Item = new Item(produit, quantite);	
    this.panier.add(item);
    }
```
je te laisse prendre en compte les autres impacts. Demande moi si tu as besoin d'aide (commit/push ton code avant que je puisse voir ce que tu aura commencé à faire).

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/view/CatalogueView.java:
```diff
@@ -0,0 +1,49 @@
+public class CatalogueView{
+    
+    private ProduitView prodV = new ProduitView();
+    public CatalogueView(CatalogueController catCont){
```
Le paramètre est inutile. Du coups le constructeur devient inutile. Supprime ce constructeur.

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/view/MenuView.java:
```diff
@@ -0,0 +1,128 @@
+			+ "\t2. Consulter la liste des produits achetés";
+		break;
+	    default:
+		this.messageErreur();
```
pour rester "consistant" avec les `case` du `switch`, mais dans la variable "message" le texte renvoyé par this.messageErreur() (en modifiant cette méthode pour quel renvoie une String, plutôt que l'afficher directement).

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/view/PanierView.java:
```diff
@@ -0,0 +1,48 @@
+
+public class PanierView{
+    
+    public PanierView(PanierController panierCont){
```
Idem que pour le constructeur de  org.iae.annecy.st1.etape1.view.CatalogueView.

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/view/PanierView.java:
```diff
@@ -0,0 +1,48 @@
+     * 		le client à qui appartient le panier
+     * @return La liste des produits du panier 
+     */
+    public String afficherPanier(Person p){
```
Évite les nom de variable trop court (3 caractère ou moins). Si tu n'a pas d'idée pour le nom de l'attribut utiliser le même nom que lue la Classe mains avec un minuscule au début est souvent correcte.

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/view/PanierView.java:
```diff
@@ -0,0 +1,48 @@
+	String text = "Réf :\t\tNom :\t\tPrix :\t\tQté :\t\tPrix total par produit :\n";
+	double total = 0;
+	for (Produit produit : p.getPanier()) {
+	    text += produit.getReference() + "\t\t" + produit.getNom() + "\t\t" + produit.getPrix() + "€\t\t" + produit.getQuantite() + "\t\t" + 
```
En Java il faut éviter le "+" avec les chaine de texte (dès que tu fait plus de concaténation sur une même chaine). Cela fonctionne bien mais est très consommateur en ressource (à cause l’immutabilité de String). Je te recommande d'utiliser un StringBufer à la place (https://docs.oracle.com/javase/7/docs/api/java/lang/StringBuffer.html). Par "convention" on appel souvent la variable du StringBuffer "sb" (ce "sb" fait partie des "cas particulier" de non de variable "court acceptable").

- Déclare ton StringBuffer avant ta boucle "for".
- A la place de chaque "+" fait un "sb.append("xxxxxxxx");
- Remplace ton return par un `return sb.toString()`.

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/view/PanierView.java:
```diff
@@ -0,0 +1,48 @@
+	double total = 0;
+	for (Produit produit : p.getPanier()) {
+	    text += produit.getReference() + "\t\t" + produit.getNom() + "\t\t" + produit.getPrix() + "€\t\t" + produit.getQuantite() + "\t\t" + 
+		    + produit.getQuantite()*produit.getPrix() + "€\n";
```
Evite le "\n" pour un saut de ligne. Cette séquence est dépendante de l'OS (la séquence change entre Windows et Linux/Mac). A la place tu peux utiliser `System.getProperty("line.separator")`.

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/view/PanierView.java:
```diff
@@ -0,0 +1,48 @@
+     * 		Le client à qui appartient la commande
+     * @return La liste des produits qu'il a commandé
+     */
+    public String afficherCommandes(Person p){
```
Même remarques que pour la méthode ci-dessus `afficherPanier(....)`

- Nom variable court
- StringBuffer
- Saut de ligne (\n).

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/view/PersonView.java:
```diff
@@ -0,0 +1,18 @@
+     * 		le client concerné
+     * @return une description du client sous forme de chaine de texte
+     */
+    public String affichageClients(Person p){
```
Mêmes remarques que pour org.iae.annecy.st1.etape1.view.afficherPanier : 

- Nom variable court
- StringBuffer

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/view/ProduitView.java:
```diff
@@ -0,0 +1,32 @@
+     * 		le produit concerné
+     * @return une chaine de texte
+     */
+    public String affichageProduit(Produit p){
```
Mêmes remarques que pour org.iae.annecy.st1.etape1.view.afficherPanier :
- Nom variable court
- StringBuffer

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/view/ProduitView.java:
```diff
@@ -0,0 +1,32 @@
+     * 		le produit concerné
+     * @return une chaine de texte
+     */
+    public String affichageProdComplet(Produit p){
```
Mêmes remarques que pour org.iae.annecy.st1.etape1.view.afficherPanier :
- Nom variable court
- StringBuffer

jderuette reviewed (commented):
Bonjour : globalement, je t'aiderais à mettre en place MVC, tes controllers ne font pas "ce qu'ils devraient". Mais d'abord un peu de "nettoyage".

Commit:
1er commit refactoring

Met a jour ta méthode
org.iae.annecy.st1.etape1.model.person.BaseDeDonnees.retrieveClient(String,
String) pour prendre ne compte le renvoie systématique d'une "Person".
==> je n'ai pas compris ce qu'il faut changer

Pour la gestion des quantités, je n'ai pas réussi à gérer le lien entre
Item et Produit quand on ajoute un produit dans le panier

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/model/produit/Produit.java:
```diff
@@ -0,0 +1,81 @@
+    private String reference, descritpion, nom, descripLongue;
+    private Double prix;
+    private Integer quantite;
+    private Item item;
```
(critique) Ta Classe "item" contient un produit. Le lien dans ce sens n'est pas "utile" et même conceptuellement faux, car semble indiquer que chaque Produit est lié à (aucun ou) UN seul Item. Hors il peut y avoir dans plusieurs paniers (de différent clients) plusieurs item avec le même produit (et des quantité différentes). Je pense que tu as fait cela pour "t'aider" dans la classe Main (ligne 568 par exemple), mais attention ton "Item" dans ton Produit n'est pas initialisé, tu as donc un "NullPointerException". Dans ta Class "Paniier" la méthode "add" à le bon comportement.
- dans Panier ajoute une methode "Item getProduit(Produit prod)" qui recherche si un Item contient ce produit. Si oui : retourne l'item de ce panier, sinon return null.
- Dans Panier, méthode "add", tu peux ajouter la logique : si j'ai déjà un Item dans ce panier (via "getItem") avec se produit, je modifie la quantité. Cela simplifiera le code de la classe Main.

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/view/PanierView.java:
```diff
@@ -0,0 +1,51 @@
+     */    
+    public String afficherPanier(Person person){
+	StringBuffer sb = new StringBuffer("Réf :\t\tNom :\t\tPrix :\t\tQté :\t\tPrix total par produit :");
+	System.getProperty("line.separator");
```
(majeur) il faut que tu utilise le StringBuffer, sinon ton saut de ligne s’affichera nullPart (sb.append(System.getProperty("line.separator"));".

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/view/PanierView.java:
```diff
@@ -0,0 +1,51 @@
+	for (Item item : person.getPanier()) {
+	    sb.append(item.getProduit().getReference()) .append("\t\t") .append(item.getProduit().getNom()) .append("\t\t") .append(item.getProduit().getPrix()) .append("€\t\t") .append(item.getQuantite()) .append("\t\t")  
+	    .append(item.getQuantite()*item.getProduit().getPrix());
+	    System.getProperty("line.separator");
```
(sb.append(System.getProperty("line.separator"));".

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/view/PanierView.java:
```diff
@@ -0,0 +1,51 @@
+     */
+    public String afficherCommandes(Person person){
+	StringBuffer sb = new StringBuffer("Réf :\t\tNom :\t\tPrix :\t\tQté :\t\tPrix total par produit");
+	System.getProperty("line.separator");
```
(sb.append(System.getProperty("line.separator"));".

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/view/PanierView.java:
```diff
@@ -0,0 +1,51 @@
+	for (Item item : person.getPanier()) {
+	    sb.append(item.getProduit().getReference()) .append("\t\t") .append(item.getProduit().getNom()) .append("\t\t") .append(item.getProduit().getPrix()) .append("€\t\t") 
+	    .append(item.getQuantite()) .append("\t\t") .append(item.getQuantite()*item.getProduit().getPrix()) .append("€");
+	    System.getProperty("line.separator");
```
(sb.append(System.getProperty("line.separator"));".

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/Main.java:
```diff
@@ -4,48 +4,738 @@
-
-	static {
-		Main.mainController = new MainController();
+	try {
```
Tu peux mettre le code (jusqu’à la ligne 71 environ) dans la Classe "Catalogue" (dans une méthode "load"). C'est bien de la responsabilité de cette Class de lire/écrire et manipuler le catalogue. Tu pourra aussi y transférer le code d'initialisation des produit (actuellement ligne 44,45,46,47)

jderuette commented on etape1/src/main/java/org/iae/annecy/st1/etape1/Main.java:
```diff
@@ -4,48 +4,738 @@
 	 */
-	public static void main(final String[] args) {
-		initUserModel();
+	try {
```
Comme pour le Catalogue, ta classe BaseDeDonnées devrait prendre en charge ce code (jusqu’à la ligne 86 environ) ainsi que l’initialisation des 2 clients par défaut. Je te propose de créer une méthode "load" pour ce code.

jderuette reviewed (changes requested):
Pense a faire "ctrl+shift+F" régulièrement pour formater le code source (ou fait un clique droit dans le code puis "source->format") pour que le code soit correctement formaté.
